st activity upload (#17)
fhdsl/UMich_ITN_Workshop: st_tools_bonus_activity.Rmd

Euclidean distances between the high score cells is calculated.

The next step involves a permutation process, in which a null distribution is
generated in order to test if the (sum of) distances among high expression cells
are smaller than expected. To that end, a random sample of cells (regardless of
expression) is selected. The random sample has the same size as the number of high
expression cells. Then, the sum of distances among the randomly sampled cells is
calculated. The random selection is repeated as many times as requested (`reps``).
Finally, a p-value is calculated by noting how many times the sum of random distances
was higher than the sum of distances among high expression cells. If the sum of
random distances was most of the times higher than the sum of distances among high
expression cells, then the null hypothesis of no spatial aggregation is rejected
(i.e., cells with high gene set expression are more aggregated than expected by chance).

The first step to run `STenrich` is to obtain a list of gene sets. Here, the `msigdbr`
package will be used to obtain HALLMARK gene sets:

```{r genesets, message=F, eval=FALSE}
# Load msigdbr

> library('msigdbr')

Reviewer: Will this need installed for some people?

should be used. Otherwise, `spatialGE` will not be able to identify genes within
the gene set.

The `STenrich` function can now be used. For this analysis, 1000 replicates (`reps=1000`)
will be used to generate the null distribution. In addition, for a gene set to be
considered as highly expressed in cell, it should be expressed at least over the
average gene set score plus 1.5 standard deviations (`num_sds=1.5`). Gene sets with
less than five genes in the sample (`min_genes=5`) and gene sets with less than 10
highly expressed cells (`min_units=10`) will not be tested. Notice a seed was set
(`seed=12345`), which critical for replication of the results in the future.

```{r stenrich, eval=FALSE}
stenrich_df <- STenrich(lc_stlist, gene_sets=gene_sets,
                        reps=1000,
                        num_sds=1.5,
                        min_genes=5,
                        min_units=10,
                        seed=12345)
```

The result of the function is a list of data frames, with one data frame per
samples (FOV in this case). For example, this is the data frame for the first
FOV in the STlist:

```{r stenrich_df, eval=FALSE}

> head(stenrich_df[['Lung5_Rep1_fov_1']])

Reviewer: This was NULL for me when I ran it

```

```{r notch_sig2, eval=FALSE}
bind_rows(stenrich_df) %>%
  mutate(prop_gene_set=size_test/size_gene_set) %>%
  filter(prop_gene_set >= 0.3 & adj_p_value < 0.05) %>%
  arrange(desc(prop_gene_set), adj_p_value) %>%
  filter(gene_set == 'HALLMARK_NOTCH_SIGNALING') %>%
  filter(str_detect(sample_name, "Lung6")) %>%
  slice_head(n=3)
```

Plots can be generated like so:

```{r notch_plots, fig.width=10, eval=FALSE}
qp <- STplot(lc_stlist, genes=gene_sets[names(gene_sets) == 'HALLMARK_NOTCH_SIGNALING'],
             color_pal='YlOrBr')

ggpubr::ggarrange(plotlist=qp, ncol=3, nrow=2, common.legend=TRUE)
```

Although expression of the NOTCH signaling pathway is higher overall in "Lung6",
the expression is not uniform, but rather is arranged in patterns within each FOV.
Conversely, cells in FOVs of "Lung5_Rep1" that show high expression of the NOTCH
signaling pathway are scattered across the tissue.

Since this expression of this pathway has been identified as spatially aggregated,
it might be of interest to know if there is correspondence of NOTCH signaling with
tissue domains. With the `STclust` function, tissue domains can be identified.

```{r stclust_smi, eval=FALSE}

> lung <- STclust(lung, ws=0.02, ks='dtc')

Reviewer: ```suggestion
lung <- STclust(lc_stlist, ws=0.02, ks='dtc')
```

created at the beggining of this tutorial will be used to generate the file paths
to the folder containng the images. Importantly, the images need to be renamed
according to the FOV name in the STlist. This renaming was already performed for
the files in the spatialGE_Data GitHub repository. Here is an example of how the
files were renamed from the original data set:

- For tissue slice "Lung5_Rep1", images were renamed from "CellComposite_F002.jpg",
"CellComposite_F006.jpg", and "CellComposite_F011.jpg" to "Lung5_Rep1_fov_2.jpg",
"Lung5_Rep1_fov_6.jpg", and "Lung5_Rep1_fov_11.jpg" respectively.
- For tissue slice "Lung6", images were renamed from "CellComposite_F004.jpg",
"CellComposite_F006.jpg", and "CellComposite_F007.jpg" to "Lung6_fov_4.jpg",
"Lung6_fov_6.jpg", and "Lung6_fov_7.jpg" respectively.

Now, upload the images to the STlist with the `load_images` function:

```{r loadimages1, eval = FALSE}
img_fp = '~/Desktop/nanostring_nsclc/Lung5_Rep1-Flat_files_and_images/CellComposite'
lc_stlist <- load_images(lc_stlist, images=img_fp)
```

> The images for slide "Lung6" were not found, which are in the corresponsing folder. The `load_images` function is ran again to add images from "Lung6":

Reviewer: ```suggestion
The images for slide "Lung6" were not found, which are in the corresponding folder. The `load_images` function is ran again to add images from "Lung6":
```

```{r loadimages2, eval = FALSE}
img_fp = '~/Desktop/nanostring_nsclc/Lung6-Flat_files_and_images/CellComposite'

> lc_stlist <- load_images(lc_stlist, images=img_fp)

Reviewer: I'm not sure if I did something wrong, but rather than 'adding" the Lung6 images to the Lung 5 images, this told me:
>Image for sample Lung5_Rep1_fov_2 was not found.
Image for sample Lung5_Rep1_fov_6 was not found.
Image for sample Lung5_Rep1_fov_11 was not found.

```

The domains can be plotted with the `STplot` function too.

```{r plot_domains, fig.width=12, eval=FALSE}
dom_p <- STplot(lung, ks='dtc', ws=0.02, deepSplit=FALSE,
                color_pal='discreterainbow')

ggpubr::ggarrange(dom_p$Lung5_Rep1_fov_2_stclust_spw0.02_dsplFalse,

>                   ti$image_Lung5_Rep1_fov_2,

Reviewer: no ti object prior to this in the code that I can find

                  dom_p$Lung5_Rep1_fov_6_stclust_spw0.02_dsplFalse,
                  ti$image_Lung5_Rep1_fov_6,
                  dom_p$Lung5_Rep1_fov_11_stclust_spw0.02_dsplFalse,
                  ti$image_Lung5_Rep1_fov_11,
                  dom_p$Lung6_fov_4_stclust_spw0.02_dsplFalse,
                  ti$image_Lung6_fov_4,
                  dom_p$Lung6_fov_6_stclust_spw0.02_dsplFalse,
                  ti$image_Lung6_fov_6,
                  dom_p$Lung6_fov_7_stclust_spw0.02_dsplFalse,
                  ti$image_Lung6_fov_7,
                  ncol=4, nrow=3)

# If no tissue images are available:
# ggpubr::ggarrange(plotlist=dom_p, ncol=3, nrow=2)
```

Now, to identify the domains, differential gene expression can be performed:

```{r domain_de, eval=FALSE}

> deg <- STdiff(lc_stlist, w=0.02, k='dtc', deepSplit=FALSE, test_type='wilcoxon')

Reviewer: Error I got here: 
>No samples left to test. Are the requested annotations/clusters present in at least one sample?

The name of the column is `stclust_spw0.02_dsplFalse` (STclust results using spatial
weight of 0.02 and DynamicTreeCuts with no deepSplit). This name is used in the
`annot` argument of `STgradient`:

```{r stgradient_run, eval=FALSE}
stg <- STgradient(lc_stlist,
                  topgenes=1000,
                  annot='stclust_spw0.02_dsplFalse',
                  ref=3,
                  samples=c('Lung6_fov_4', 'Lung6_fov_6', 'Lung6_fov_7'),
                  distsumm='min',
                  robust=TRUE, cores=2)
```

Here are the first few rows of the `STgradient` result for sample "Lung6_fov_4":

```{r stg_res, eval=FALSE}
head(stg[['Lung6_fov_4']])
```

The key columns to look at are `min_spearman_r` and `min_spearman_r_pval_adj`.
Now, to find genes with significant spatial gradients across the three samples,
these commands can be used:

```{r topstg, eval=FALSE}
bind_rows(stg) %>%
  arrange(desc(abs(min_spearman_r))) %>%
  group_by(sample_name) %>%
  slice_head(n=5) %>%
  ungroup() %>%
  arrange(gene)
```

It can be seen from these results that multiple keratin genes show spatial gradients,
all of them with negative Spearman's coefficients (cells close to domain 3 tend to

> have higher keratin expression). Other such __HLA-B__ and __NDRG1__ have higher

Reviewer: I don't see NDRG1 in the table above when I run it

expression in cells far from domain 3:

```{r stg_plots, eval=FALSE}
qp2 <- STplot(lc_stlist, genes=c('KRT6A', 'HLA-B'), samples='Lung6_fov_4', color_pal='YlOrBr')
ggpubr::ggarrange(plotlist=qp2, common.legend=TRUE)
```

Here are the results for genes within the NOTCH signaling pathway:

```{r stg_notch, eval=FALSE}
bind_rows(stg) %>%
  arrange(desc(abs(min_spearman_r))) %>%
  filter(gene %in% gene_sets[['HALLMARK_NOTCH_SIGNALING']]) %>%
  filter(min_spearman_r_pval_adj < 0.05)
```

Although the correlations in genes from this pathway are not as large as and those
for __KRT6A__ and __HLA-B__ (i.e., lower absolute Spearman's coefficient), it can

> be seen that the genes __NOTCH1__ and __NOTCH3__ show evidence of expression

Reviewer: Also don't see NOTCH3 in the table above when I run it

---
title: "Bonus activity: Using the command-line for single-cell Spatial Transcriptomics analysis"
output:
  html_document:
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
css: style.css
editor_options:
  markdown:
    wrap: 72
---

```{r echo = FALSE, show = FALSE}
knitr::opts_chunk$set(number_sections = FALSE)
```

> 

Reviewer 2: ```suggestion
## Accessing R and RStudio
```

To continue to practice using `spatialGE` using R, you can access our PositCloud instance until April 27th. See the [clinical workshop activity guidance](https://hutchdatascience.org/UMich_ITN_Workshop/clinical_data_activity.html) for instructions on how to connect.

> Alternatively you can download and install R and RStudio to use on your local machine or see if your institution has a sever available with R and R Studio.

Reviewer 2: Alternatively you can download and install R and RStudio to use on your local machine or see if your institution has a sever available with R and RStudio. See this [link for information on installations](https://jhudatascience.org/intro_to_r/modules/Setup/Setup.html). You could also consider using another cloud-based option like [AnVIL](https://jhudatascience.org/AnVIL_Book_Getting_Started/starting-rstudio.html) or [Galaxy](https://training.galaxyproject.org/training-material/topics/galaxy-interface/tutorials/rstudio/tutorial.html).

> Alternatively you can download and install R and RStudio to use on your local machine or see if your institution has a sever available with R and R Studio.

Reviewer 2: ```suggestion
 Alternatively you can download and install R and RStudio to use on your local machine or see if your institution has a sever available with R and RStudio. See this [link for information on installations](https://jhudatascience.org/intro_to_r/modules/Setup/Setup.html). You could also consider using another cloud-based option like [AnVIL](https://jhudatascience.org/AnVIL_Book_Getting_Started/starting-rstudio.html) or [Galaxy](https://training.galaxyproject.org/training-material/topics/galaxy-interface/tutorials/rstudio/tutorial.html).
```

Reviewer 2: @oospina What do you think of this?

> See this [link for information on installations](https://jhudatascience.org/intro_to_r/modules/Setup/Setup.html). 

Reviewer 2: ```suggestion
## `spatialGE` R package
```
